app/gui2/src/util/shortcuts.ts

```diff
@@ -0,0 +1,400 @@
+    }
+  }
+
+  function keyboardHandler(
+    handlers: Partial<Record<BindingName | typeof DefaultHandler, (event: KeyboardEvent) => void>>,
+    stopImmediatePropagationOnMatch = true,
+    preventDefaultOnMatch = true,
+  ): (event: KeyboardEvent) => boolean {
+    return (event) => {
+      const eventModifierFlags = modifierFlagsForEvent(event)
+      const keybinds = keyboardShortcuts[event.key.toLowerCase() as Key_]?.[eventModifierFlags]
+      let handle = handlers[DefaultHandler]
+      if (keybinds != null) {
+        for (const bindingName in handlers) {
+          if (keybinds?.has(bindingName as BindingName)) {
+            handle = handlers[bindingName as BindingName]
+            break
+          }
+        }
+      }
+      if (handle == null) {
+        return false
+      }
+      if (stopImmediatePropagationOnMatch) {
+        event.stopImmediatePropagation()
+      }
+      if (preventDefaultOnMatch) {
+        event.preventDefault()
+      }
+      handle(event)
+      return true
+    }
+  }
+
+  function mouseHandler(
+    handlers: Partial<Record<BindingName | typeof DefaultHandler, (event: MouseEvent) => void>>,
+    stopImmediatePropagationOnMatch = true,
+    preventDefaultOnMatch = true,
+  ): (event: MouseEvent) => boolean {
+    return (event) => {
+      const eventModifierFlags = modifierFlagsForEvent(event)
+      const keybinds = mouseShortcuts[event.buttons as PointerButtonFlags]?.[eventModifierFlags]
+      let handle = handlers[DefaultHandler]
+      if (keybinds != null) {
+        for (const bindingName in handlers) {
+          if (keybinds?.has(bindingName as BindingName)) {
+            handle = handlers[bindingName as BindingName]
+            break
+          }
+        }
+      }
+      if (handle == null) {
+        return false
+      }
+      if (stopImmediatePropagationOnMatch) {
+        event.stopImmediatePropagation()
+      }
+      if (preventDefaultOnMatch) {
+        event.preventDefault()
+      }
+      handle(event)
+      return true
+    }
+  }
+
+  return { keyboardHandler, mouseHandler }
+}
+
+/** A type predicate that narrows the potential child of the array. */
+function includesPredicate<T extends U, U>(array: readonly T[]) {
+  const array_: readonly unknown[] = array
+  return (element: unknown): element is T => array_.includes(element)
+}
+
+const isModifier = includesPredicate(allModifiers)
+const isPointer = includesPredicate(allPointers)
+
+function decomposeKeybindString(string: string): ModifierStringDecomposition {
+  const parts = string
+    .trim()
+    .split(/[\s+]+/)
+    .map((part) => normalizedKeyboardSegmentLookup[part] ?? part)
+  const modifiers = parts.filter(isModifier)
+  const key = parts.find((part) => !isModifier(part))
+  return {
+    // If the key is blank, assume it was a `+` that was removed by the splitting regex.
+    key: key === '' ? '+' : key ?? '',
+    modifiers,
+  }
+}
+
+function parseKeybindString(string: string): Keybind | Mousebind {
```

Comment:
This could have a few unit tests. Those can be added at the end of this file inside an `if (import.meta.vitest)` block.